File: src/main/java/com/distributedappspu/cscources/services/CourseService.java
package com.distributedappspu.cscources.services;

import com.distributedappspu.cscources.mapper.CourseMapper;
import com.distributedappspu.cscources.models.dto.CourseDTO;
import com.distributedappspu.cscources.models.entities.CourseEntity;
import com.distributedappspu.cscources.models.entities.InstructorEntity;
import com.distributedappspu.cscources.repositories.CourseRepository;
import com.distributedappspu.cscources.repositories.InstructorRepository;
import jakarta.validation.Valid;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.Date;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Service
public class CourseService {

    private final CourseRepository courseRepository;

    private final InstructorRepository instructorRepository;

    private final CourseMapper courseMapper;

    public CourseService(CourseRepository courseRepository, CourseMapper courseMapper, InstructorRepository instructorRepository) {
        this.courseRepository = courseRepository;
        this.courseMapper = courseMapper;
        this.instructorRepository = instructorRepository;
    }

    public List<CourseDTO> getAllCourses() {
        return courseMapper.mapCourses(this.courseRepository.findAll());
    }

    public CourseDTO getCourseById(UUID id) {
        return courseMapper.mapCourse(courseRepository.findById(id).orElseThrow());
    }

    public Page<CourseDTO> getCoursesByName(String name, Pageable pageable) {
        Page<CourseEntity> courseEntityPage = courseRepository.findCourseEntitiesByName(name, pageable);
        return courseEntityPage.map(courseMapper::mapCourse);
    }

    public Page<CourseDTO> getCoursesByStartDate(Date startDate, Pageable pageable) {
        Page<CourseEntity> courseEntityPage = courseRepository.findCourseEntitiesByStartDate(startDate, pageable);
        return courseEntityPage.map(courseMapper::mapCourse);
    }

    public Page<CourseDTO> getCoursesByEndDate(Date endDate, Pageable pageable) {
        Page<CourseEntity> courseEntityPage = courseRepository.findCourseEntitiesByEndDate(endDate, pageable);
        return courseEntityPage.map(courseMapper::mapCourse);
    }

    public Page<CourseDTO> getCoursesByInstructorId(UUID instructorId, Pageable pageable) {
        Optional<InstructorEntity> instructor = instructorRepository.findById(instructorId);
        if (instructor.isEmpty()) {
            throw new IllegalArgumentException("Instructor with ID " + instructorId + " not found.");
        }
        Page<CourseEntity> courseEntityPage = courseRepository.findCourseEntitiesByInstructor(instructor.get(), pageable);
        return courseEntityPage.map(courseMapper::mapCourse);
    }

    public CourseDTO createCourse(@Valid CourseDTO courseDTO){
        return addCourse(courseDTO);
    }

    public CourseDTO updateCourse(UUID id, @Valid CourseDTO courseDTO) {

        CourseEntity existingCourse = courseRepository.findById(id).orElse(null);
        if (existingCourse == null) {
            throw new IllegalArgumentException("Course does not exist!");
        }
        return addCourse(courseDTO);
    }

    private CourseDTO addCourse(@Valid CourseDTO courseDTO) {
        CourseEntity courseEntity = courseMapper.mapCourse(courseDTO);
        InstructorEntity instructor = instructorRepository.findById(courseDTO.getInstructorId()).orElse(null);
        courseEntity.setInstructor(instructor);
        CourseEntity newCourse = courseRepository.save(courseEntity);
        CourseDTO newCourseDTO = courseMapper.mapCourse(newCourse);
        newCourseDTO.setInstructorId(courseEntity.getInstructor().getId());
        return newCourseDTO;
    }

    public void deleteCourse(UUID id) {
        if(!courseRepository.existsById(id)) {
            return;
        }
        courseRepository.deleteById(id);
    }
}

File: cs-cources-ui/app/services/courseService.ts
import axios from 'axios';
import { CourseDTO, Page } from '../types';

const API_BASE_URL = 'http://localhost:8080/v1/courses';

const apiClient = axios.create({
    baseURL: API_BASE_URL,
});

export const createCourse = (courseDTO: CourseDTO) => {
    return apiClient.post<CourseDTO>('/createCourse', courseDTO);
};

export const updateCourse = (id: string, courseDTO: CourseDTO) => {
    return apiClient.put<CourseDTO>(`/${id}`, courseDTO);
};

export const getCourseById = (id: string) => {
    return apiClient.get<CourseDTO>(`/${id}`);
};

export const getCoursesByName = (name: string, pageable: any) => {
    // Implement pageable or pagination as needed
    return apiClient.get<Page<CourseDTO>>(`/byName?name=${name}`, { params: pageable });
};

export const getCoursesByStartDate = (startDate: string, pageable: any) => {
    return apiClient.get<Page<CourseDTO>>(`/byStartDate?startDate=${startDate}`, { params: pageable });
};

export const getCoursesByEndDate = (endDate: string, pageable: any) => {
    return apiClient.get<Page<CourseDTO>>(`/byEndDate?endDate=${endDate}`, { params: pageable });
};

export const getCoursesByInstructorId = (instructorId: string, pageable: any) => {
    return apiClient.get<Page<CourseDTO>>(`/byInstructor/${instructorId}`, { params: pageable });
};

export const getAllCourses = () => {
    return apiClient.get<CourseDTO[]>(`/getAllCourses`);
};

export const deleteCourse = (id: string) => {
    return apiClient.delete<void>(`/${id}`);
};
